### DISCUSSION TITLE
Is it possible to read a solution file created using highspy in one format or style, and convert it to another format without the need to run the model?

### DISCUSSION THREAD

**Discussion by hbshrestha** [ERGO-Code/HiGHS#2075] [Q&A] [answered]
Hello. I ran a model and created a solution file using style 1 with the following code a week ago. 
```
import highspy

h = highspy.Highs()
filename = 'abc.mps'
h.readModel(filename)

h.setOptionValue("solver","ipm")
h.run()

h.writeSolution("abc_highs_solution1.sol", 1)
```
I learnt yesterday #2072 that it is also possible to write the solution file in different format using HIGHS. As per the [documentation](https://ergo-code.github.io/HiGHS/dev/options/definitions/#write_solution_style), style 1 refers to HIGHS pretty format, and style 3 refers to Glpsol pretty format. 
![image](https://github.com/user-attachments/assets/cde87569-b09d-43c3-ba3d-75104300af3b)

I need the solution file in style 3. The model had taken two days to run. Therefore, I'd like to avoid running the model again if at all possible. 1. Is it possible to read the `abc_highs_solution1.sol` file using highspy and convert it into style 3 directly without requiring to run the model again? 

I have a second sub-question.
2. In the style 3 solution file, can the index number go beyond six digits (to 1 million and further)?

**Comment by jajhall** [comment]
This should work.

If the problem is an LP, call run() and HiGHS should just validate that the solution read in is optimal, and then write out in Glpsol format. 

If it's a MIP, it's less clear. Don't call run(), but hope that it writes out the solution, despite not having the model status set.

It would have been sensible to check that this mechanism works before solving such a large problem, but I know life's not like that at times!

**Reply by hbshrestha** [reply]
Thank you for the response! It is an LP problem.

I noticed it is possible to read solution. When I tried, 

```
h = highspy.Highs()
filename = "abc_highs_solution1.sol"
h.readsolution("abc_highs_solution1.sol", 1)
```
I got `<HighsStatus.kError: -1>`. 

One has to pass arg0 and arg1. But, what do they refer to?
![image](https://github.com/user-attachments/assets/63d4582d-9816-433c-b0b9-5f239dc4b47e)

**Reply by hbshrestha** [reply]
Thank you. Please take your time.

**Comment by jajhall** [comment]
Unfortunately - and I've looked back properly at your earlier questions 

> As per the [documentation](https://ergo-code.github.io/HiGHS/dev/options/definitions/#write_solution_style), style 1 refers to HIGHS old raw format, and style 3 refers to Glpsol pretty format.

 `write_solution_style = 1` is not a HIGHS raw format (-1 => old raw; 0 => current raw), but HiGHS "pretty" format. HiGHS can't read its pretty format, as this is for humans. 

So, you'll have to convert the HiGHS "pretty" format file into Glpsol "pretty" format using some text processing package (like perl/awk or some modern equivalent)

> 2. In the style 3 solution file, can the index number go beyond six digits (to 1 million and further)?

Yes, up to 9 digits!

In passing, there is at least one bug in the code to write out in Glpsol "pretty" format. I get lines like 

>    No.   Row name   St   Activity     Lower bound   Upper bound    Marginal
------ ------------ -- ------------- ------------- ------------- -------------
     1 .Z....       B         225495                             
     2 ....01       NU             0                           0         -3310 
     3 ....02       NS          52.6          52.6             =       775.224 

I can look into this and probably soon fix it

**Reply by hbshrestha** [reply]
Thank you for the response! Could you highlight what is the bug here that you are referring to?
![image](https://github.com/user-attachments/assets/e217d347-5bce-4ba0-9e3b-3f9d5cd3ef88)

Is it the equal to sign (=) in Upper bound column or something else?

**Reply by jajhall** [reply]
Yes, and I get other odd values in later columns

**Reply by hbshrestha** [reply]
Indeed, it'd be good to have them fixed. May I ask how long it might take to fix this?

**Reply by hbshrestha** [reply]
Besides, the = sign, I also noticed "<eps" mentioned in Marginal column. I am not sure what it means and if it is a bug.
![image](https://github.com/user-attachments/assets/a13c8f71-0c38-48d3-a9ef-5c7f25fff15e)

**Reply by jajhall** [reply]
Yes, I saw that, too. I think I know the cause, and it should be simple to fix.

**Reply by jajhall** [reply]
> Besides, the = sign, I also noticed "<eps" mentioned in Marginal column. I am not sure what it means and if it is a bug.

Looking at the code, writing "=" for the upper bound when it's equal to the lower bound, and "< eps" for the "Marginal" when it is less than 1e-9, is clearly hard-coded, so I have to assume that it's the correct format for the "pretty" Glpsol file. When I wrote the Glpsol output utility, I was studying the Glpsol source code to see how it handled these cases.

**Reply by hbshrestha** [reply]
As per your response, I'd understand there is not a bug. "=" is just a different way of denoting same value as bounds, and "<eps" would represent less than 1e-9. So, we can still use the solution file, directly, right?

**Reply by jajhall** [reply, marked as answer]
Yes, I correctly reproduced the glpsol style in the pretty format

**Reply by hbshrestha** [reply]
Thank you.

**Comment by jajhall** [comment]
> By passing 3, I am getting very similar to the GLPK format I shared in the question. Header names are different.
h.writeSolution("abc_highs_solution.sol", 3)

The format I created was for `Glpsol`, not `GLPK`. I know that they are closely related, and if the header files are different, I can try to add `GLPK` raw and pretty format - as I'm keen to wean people off `Glpsol`/`GLPK` since their solvers have such poor performance.

Can `Glpsol`/`GLPK` read their own pretty format?

**Reply by hbshrestha** [reply]
I could still use the style 3 solution file for getting results in my small model tests.
I am not sure whether Glpsol/GLPK can read their own pretty format.
